src/impression.js

```diff
@@ -47,15 +75,57 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) We need a timeout here?
+    // TODO(@zhouyx) need test with a real response.
+    invoke(win, dev().assertString(clickUrl)).then(response => {
+      applyResponse(win, viewer, response);
+      resolveImpression();
+    });
   });
 }
 
+/**
+ * Send the url to ad server and wait for its response
+ * @param {!Window} win
+ * @param {string} clickUrl
+ * @return {!Promise<!JSONType>}
+ */
 function invoke(win, clickUrl) {
-  xhrFor(win).fetchJson(clickUrl, {
+  if (getMode().localDev && !getMode().test) {
+    clickUrl = 'http://localhost:8000/impression-proxy?url=' + clickUrl;
+  }
+  return xhrFor(win).fetchJson(clickUrl, {
     credentials: 'include',
     requireAmpResponseSourceOrigin: true,
   });
-  // TODO(@cramforce): Do something with the result.
+}
+
+/**
+ * parse the response back from ad server
+ * Set for analytics purposes
+ * @param {!Window} win
+ * @param {!Object} response
+ */
+function applyResponse(win, viewer, response) {
+  const adLocation = response['location'];
+  const adTracking = response['tracking'];
+
+  // If there is a tracking_url, need to track it
+  // Otherwise track the location
+  const trackUrl = adTracking || adLocation;
+
+  if (trackUrl && !isProxyOrigin(trackUrl)) {
+    new Image().src = trackUrl;
+  }
+
+  // Replace the location href params with new location params we get.
+  if (adLocation) {
+    const currentHref = win.location.href;
+    const url = parseUrl(adLocation);
+    const params = parseQueryString(url.search);
+    const newHref = addParamsToUrl(currentHref, params);
+    win.history.replaceState(null, '', newHref);
```

Comment:
@lannka persuaded me that blindly replacing URL is the right way to do here on my bus home yesterday 😄 . It will help to keep behaviors same with what we have now w/o ALP. Server most likely will keep all query params and append `gclid` to the end. If server really want to delete any query params that we have, we should still enable them to do that, instead of keeping it. 
@cramforce What's your idea about this?